File: src/main/java/com/raychenon/leetcode/LongestPalindromicSubstring.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 2019-09-29
 * https://leetcode.com/problems/longest-palindromic-substring/
 */
public class LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) return "";

        int start = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expendAroundCenter(s, i, i);
            int len2 = expendAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }

    int expendAroundCenter(String s, int start, int end) {
        int left = start, right = end;
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/LongestPalindromicSubstring.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-09-29
 * https://leetcode.com/problems/longest-palindromic-substring/
 */
object LongestPalindromicSubstring {


    /**
     * time complexity: O(n^2). Single for loop + while loop in {@link #expandAroundCenter}
     * space complexity: O(1)
     *
     * @param s
     * @return longest palindrome
     */
    fun longestPalindrome(s: String): String {
        if (s.isEmpty()) return ""

        var start = 0
        var end = 0
        for (i in 0 until s.length) {
            // max length between
            // - palindrome with odd length: [ |i| ]
            // - palindrome with even length: [ |i|i+1| ]
            val len = Math.max(lengthFromTheCenter(s, i, i), lengthFromTheCenter(s, i, i + 1))
            if (len > end - start) {
                start = i - (len - 1) / 2
                end = i + len / 2
            }
        }
        return s.substring(start, end + 1)
    }

    /**
     * \ \start\ \ \i\ \ \ \end|
     *          <---  ----->
     *          |--length--|
     */
    private fun lengthFromTheCenter(s: String, start: Int, end: Int): Int {
        var left = start
        var right = end
        while (left >= 0 && right < s.length && s[left] == s[right]) {
            left--
            right++
        }
        return right - left - 1
    }
}